samza-core/src/main/java/org/apache/samza/storage/StorageRecovery.java

```diff
@@ -229,7 +229,7 @@ private void getContainerStorageManagers() {
         .getCheckpointManager(new MetricsRegistryMap()).orElse(null);
 
     for (ContainerModel containerModel : containers.values()) {
-      ContainerContext containerContext = new ContainerContextImpl(containerModel, new MetricsRegistryMap());
+      ContainerContext containerContext = new ContainerContextImpl(containerModel, new MetricsRegistryMap(), null);
```

Comment:
minor: use Optional instead of null passing ?

Reply:
`ContainerThreadPool` when being null has special meaning in samza today and hence it is important keep that semantic unchanged for the framework components.

Additionally, use of `Optional` constructor arguments is an anti-pattern. Ideally, we could have two constructors but the usage seemed minimal to warrant two constructors.

More information on why using Optional for arguments/constructor params [1][2]

[1] http://dolszewski.com/java/java-8-optional-use-cases/
[2] https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments